plugins/obs-ipc/interface.cpp

```diff
@@ -0,0 +1,837 @@
+{
+	IPCInterface *ipc = (IPCInterface *)data;
+	signal_handler_t *handler = obs_get_signal_handler();
+
+	if (signalInstances.find(signal) == signalInstances.end()) {
+		signalInstances[signal] = 1;
+	} else {
+		signalInstances[signal]++;
+		if (signalInstances[signal] != 1) {
+			return true;
+		}
+	}
+
+	if (signal == "source_create") {
+		signal_handler_connect(handler, signal.c_str(),
+				       ipc->source_create, ipc);
+		return true;
+	} else if (signal == "source_remove") {
+		signal_handler_connect(handler, signal.c_str(),
+				       ipc->source_remove, ipc);
+		return true;
+	} else if (signal == "source_activate") {
+		signal_handler_connect(handler, signal.c_str(),
+				       ipc->source_activate, ipc);
+		return true;
+	} else if (signal == "source_deactivate") {
+		signal_handler_connect(handler, signal.c_str(),
+				       ipc->source_deactivate, ipc);
+		return true;
+	} else if (signal == "source_show") {
+		signal_handler_connect(handler, signal.c_str(),
+				       ipc->source_show, ipc);
+		return true;
+	} else if (signal == "source_hide") {
+		signal_handler_connect(handler, signal.c_str(),
+				       ipc->source_hide, ipc);
+		return true;
+	} else if (signal == "source_rename") {
+		signal_handler_connect(handler, signal.c_str(),
+				       ipc->source_rename, ipc);
+		return true;
+	} else if (signal == "source_volume") {
+		signal_handler_connect(handler, signal.c_str(),
+				       ipc->source_volume, ipc);
+		return true;
+	} else if (signal == "source_mute") {
+		signal_handler_connect(ipc->signal_handler, signal.c_str(),
+				       ipc->source_mute, ipc);
+		return true;
+	} else if (signal == "source_unmute") {
+		signal_handler_connect(ipc->signal_handler, signal.c_str(),
+				       ipc->source_unmute, ipc);
+		return true;
+	} else if (signal == "source_audio_activate") {
+		signal_handler_connect(ipc->signal_handler, signal.c_str(),
+				       ipc->source_audio_activate, ipc);
+		return true;
+	} else if (signal == "source_audio_deactivate") {
+		signal_handler_connect(ipc->signal_handler, signal.c_str(),
+				       ipc->source_audio_deactivate, ipc);
+		return true;
+	}
+
+	return false;
+}
+
+static Json CreateResponse(string command, IPCError result,
+			   Json response = nullptr)
+{
+	Json::object res =
+		Json::object{{"result", (int)result}, {"command", command}};
+
+	if (response != nullptr)
+		res["response"] = response;
+
+	return res;
+}
+
+static string GetDeviceType(string id)
+{
+	if (id == INPUT_AUDIO_SOURCE)
+		return "AudioInput";
+	if (id == OUTPUT_AUDIO_SOURCE)
+		return "AudioOutput";
+	if (id == "dshow_input")
+		return "AudioInput";
+
+	return "AudioOutput";
+}
+
+void IPCInterface::CommandResult(string command, Json params,
+				 CommandResponse callback)
+{
+	Json success = CreateResponse(command, IPCError::IPC_ERROR_SUCCESS);
+
+	if (command == "StartStreaming") {
+		obs_frontend_streaming_start();
+		callback(this, success);
```

Comment:
Might be a bit premature to report success here; see e.g. https://github.com/obsproject/obs-studio/blob/81e4211f64c7c2e8c8d1043f63fa1df2cf16dc89/UI/window-basic-main-outputs.cpp#L48-L56 + context

Reply:
I will certainly see if I can make that work without increasing the complexity of the code too much.